external/model-preparation-algorithm/mpa_tasks/apis/classification/task.py

```diff
@@ -109,19 +98,52 @@ def infer(self,
 
         dataset_size = len(dataset)
         for i, (dataset_item, prediction_item) in enumerate(zip(dataset, predictions)):
-            label = []
+            item_labels = []
+            pos_thr = 0.5
+
             if self._multilabel:
-                pos_thr = 0.5
+                if max(prediction_item) < pos_thr:
+                    logger.info('Confidence is smaller than pos_thr, empty_label will be appended to item_labels.')
+                    item_labels.append(ScoredLabel(self._empty_label, probability=1.))
+                    continue
                 for cls_idx, pred_item in enumerate(prediction_item):
                     if pred_item > pos_thr:
                         cls_label = ScoredLabel(self.labels[cls_idx], probability=float(pred_item))
-                        label.append(cls_label)
+                        item_labels.append(cls_label)
+
+            elif self._hierarchical:
+                for i in range(self._hierarchical_info['num_multiclass_heads']):
+                    logits_begin, logits_end = self._hierarchical_info['head_idx_to_logits_range'][i]
+                    head_logits = prediction_item[logits_begin : logits_end]
+                    j = np.argmax(head_logits)  # Assume logits already passed softmax
+                    label_str = self._hierarchical_info['all_groups'][i][j]
+                    ote_label = next(x for x in self._labels if x.name == label_str)
+                    item_labels.append(ScoredLabel(label=ote_label, probability=float(head_logits[j])))
+
+                if self._hierarchical_info['num_multilabel_classes']:
+                    logits_begin, logits_end = self._hierarchical_info['num_single_label_classes'], -1
+                    head_logits = prediction_item[logits_begin : logits_end]
+                    for i in range(head_logits.shape[0]):
+                        if head_logits[i] > pos_thr:  # Assume logits already passed sigmoid
+                            label_str = self._hierarchical_info['all_groups'][self._hierarchical_info['num_multiclass_heads']+i][0]
+                            ote_label = next(x for x in self._labels if x.name == label_str)
+                            item_labels.append(ScoredLabel(label=ote_label, probability=float(head_logits[i])))
+                item_labels = self._task_environment.label_schema.resolve_labels_probabilistic(item_labels)
+                if not item_labels:
+                    logger.info('item_labels is empty.')
+                    item_labels.append(ScoredLabel(self._empty_label, probability=1.))
             else:
-                label_idx = prediction_item.argmax()
-                cls_label = ScoredLabel(self._labels[label_idx], probability=float(prediction_item[label_idx]))
-                label.append(cls_label)
-            dataset_item.append_labels(label)
+                if any(np.isnan(prediction_item)):
```

Comment:
Isn't this supposed to be an 'elif' in this code-flow?

Reply:
Revised. Thanks.